ext/curl/multi.c

```diff
@@ -187,7 +187,12 @@ PHP_FUNCTION(curl_multi_select)
 
 	mh = Z_CURL_MULTI_P(z_mh);
 
-	error = curl_multi_wait(mh->multi, NULL, 0, (unsigned long) (timeout * 1000.0), &numfds);
+	if (!(timeout >= ((double)INT_MIN / 1000.0) && timeout <= ((double)INT_MAX / 1000.0))) {
+		php_error_docref(NULL, E_WARNING, "timeout must be between %d and %d", (INT_MIN / 1000), (INT_MAX / 1000));
```

Comment:
This appears to be slightly off due to rounding. Should probably be something like `(int) fceil((double) INT_MAX/1000`)`, but maybe that's overkill.

And I think [`zend_argument_value_error()`](https://heap.space/s?refs=zend_argument_value_error&project=php-src) could be used here, although that wouldn't show the docref. Maybe something that should generally be addressed (unless there is already something for this).

Reply:
Oops! Of course, throwing a `ValueError` would be something completely different, but still this may be an option for `master`, and generally, `zend_argument_value_error()` doesn't look right, since it hasn't the docref. @Girgias and @kocsismate may have something to say about that.

Reply:
It might make sense to add support for docrefs for all argument exceptions and ZPP failures.
But the docref param is only useful for INI settings, and I don't think we throw exceptions for INI settings (yet?)

Reply:
> It might make sense to add support for docrefs for all argument exceptions and ZPP failures.

There is a problem, though, since ZPP is part of the Zend engine, and `php_error_docref()` is part of PHP, and the engine shouldn't rely on PHP. I guess that is the reason why `zend_error()` has been used for this.

> But the docref param is only useful for INI settings, […]

Yeah, but I didn't mean the docref param, but was referring to the general mechanism, which is nice, e.g. https://3v4l.org/deO0B.

Reply:
Ah right... I agree that `Zend` shouldn't rely on `main/` (although I recently learnt that `Zend` does require `ext/standard`/`ext/hash` :|)

Maybe we should introduce a function pointer API similar to `zend_autoload` ?

Reply:
Oh, already forgot about the `Zend -> ext/hash` dependency. We should fix this (keeping BC seems to be possible, on the first glance).

And yes, a function pointer could solve the issue nicely.

And we should probably use some tooling to detect such circular dependencies.